### DISCUSSION TITLE
tokio use more memory than a normal go application

### DISCUSSION THREAD

**Discussion by dylankyc** [tokio-rs/tokio#3563] [Q&A]
**Description**

Hi,

I'm writing an pulsar consumer, which will consume messages from the pulsar topic. But it seems the process is using more memory and cpu than an go consumer. Why rust(a language which has no GC) consumes more memory than go(a language which has GC), which does the same thing as the rust application?

I also did a flamegraph but didn't know how to investigate how to make it consume less cpu and memory.

[flamegraph.svg.zip](https://github.com/tokio-rs/tokio/files/6048275/flamegraph.svg.zip)

Cargo project for consuming pulsar messages: https://github.com/lichuan6/pulsar-consumer
Go project to produce & consume messages: https://github.com/lichuan6/pulsar-simple-producer
Pulsar setup: 
```
wget -c "https://www.apache.org/dyn/mirrors/mirrors.cgi?action=download&filename=pulsar/pulsar-2.7.0/apache-pulsar-2.7.0-bin.tar.gz" -O apache-pulsar-2.7.0-bin.tar.gz && tar xf apache-pulsar-2.7.0-bin.tar.gz && cd apache-pulsar-2.7.0 && bin/pulsar standalone
```

Thanks

**Version**
```
│   ├── tokio v0.2.25
│   │   └── tokio-macros v0.2.6 (proc-macro)
│   ├── tokio-native-tls v0.1.0
│   │   └── tokio v0.2.25 (*)
│   ├── tokio-util v0.3.1
│   │   └── tokio v0.2.25 (*)
└── tokio v0.2.25 (*)
```

**Platform**
```
Darwin x.local 19.5.0 Darwin Kernel Version 19.5.0: Tue May 26 20:41:44 PDT 2020; root:xnu-6153.121.2~2/RELEASE_X86_64 x86_64
```

**Comment by Darksonn** [comment]
How much memory are we talking here?

**Comment by dylankyc** [comment]
<img width="1377" alt="Screen Shot 2021-02-27 at 3 43 16 PM" src="https://user-images.githubusercontent.com/74223747/109381177-a5ec5300-7913-11eb-926a-49747c284ba2.png">


@Darksonn Thanks for your reply.

Rust consumer uses about 45M memory while go consumer uses about 8M, and cpu usages is more high for rust than go in htop.

**Comment by carllerche** [comment]
I'm going to convert this into a discussion. If it turns out this exposes a bug in Tokio (as opposed to an issue in the non-tokio code), we can open a new issue & reference the discussion :+1:.

**Reply by carllerche** [reply]
It will be easier to evaluate if this is a problem w/ Tokio or not w/ a minimal repro comparing a small tokio app w/ an equivalent one w/ go.
